File: pipeline/kite9-pipeline-common/src/main/java/com/kite9/k9server/command/xml/replace/AbstractReplaceCommand.java
package com.kite9.k9server.command.xml.replace;

import com.kite9.k9server.adl.holder.pipeline.ADLDom;
import com.kite9.k9server.command.CommandException;
import com.kite9.k9server.command.xml.AbstractADLCommand;

public abstract class AbstractReplaceCommand<E, T> extends AbstractADLCommand {

	public String fragmentId;
	public String from, to;
	
	protected abstract T getFromContent(ADLDom context);
	
	protected abstract T getToContent(ADLDom context);
	
	protected abstract E getExistingContent(ADLDom in);
	
	@Override
	public Mismatch applyCommand(ADLDom in) throws CommandException {
		checkProperties();
		E existing = getExistingContent(in);
		T fromContent = getFromContent(in);
		
		Mismatch m= same(existing, fromContent);
		if (m != null) {
			return m;
		}
	
		doReplace(in, existing, getToContent(in), fromContent);
		return null;
	}

	protected void checkProperties() {
		ensureNotNull("fragmentId", fragmentId);
	}

	protected abstract void doReplace(ADLDom on, E site, T toContent, T fromContent);

	protected Mismatch same(E existing, T with) {
		if (existing.equals(with)) {
			return null;
		} else {
			return () -> "Diagram has changed since move command issued.  Expected "+with+" got "+existing;
		}
	}

	@Override
	public Mismatch undoCommand(ADLDom in) throws CommandException {
		checkProperties();
		E existing = getExistingContent(in);
		T toContent = getToContent(in);

		Mismatch m = same(existing, toContent);
		
		if (m != null) {
			return m;
		} 
		
		
		doReplace(in, existing, getFromContent(in), toContent);
		return null;
	}
	
	
}

File: pipeline/kite9-pipeline-common/src/main/kotlin/com/kite9/pipeline/command/xml/replace/AbstractReplaceCommand.kt
package com.kite9.pipeline.command.xml.replace

import com.kite9.pipeline.adl.holder.pipeline.ADLDom
import com.kite9.pipeline.command.Command.Mismatch
import com.kite9.pipeline.command.CommandContext
import com.kite9.pipeline.command.xml.AbstractADLCommand

abstract class AbstractReplaceCommand<E, T> : AbstractADLCommand() {

    var fragmentId: String = ""
    var from: String? = null
    var to: String? = null

    protected abstract fun getFromContent(context: ADLDom, ctx: CommandContext): T
    protected abstract fun getToContent(context: ADLDom, ctx: CommandContext): T
    protected abstract fun getExistingContent(o: ADLDom, ctx: CommandContext): E

    override fun applyCommand(d: ADLDom, ctx: CommandContext): Mismatch? {
        checkProperties()
        val existing = getExistingContent(d, ctx)
        val fromContent = getFromContent(d, ctx)
        val m = same(existing, fromContent)
        if (m != null) {
            return m
        }
        doReplace(d, existing, getToContent(d, ctx), fromContent, ctx)
        return null
    }

    protected open fun checkProperties() {
        ensureNotNull("fragmentId", fragmentId)
    }

    protected abstract fun doReplace(on: ADLDom, site: E, toContent: T, fromContent: T, ctx: CommandContext)

    protected open fun same(existing: E, with: T): Mismatch? {
        return if (existing == with) {
            null
        } else {
            Mismatch { "Diagram has changed since move command issued.  Expected $with got $existing" }
        }
    }

    override fun undoCommand(d: ADLDom, ctx: CommandContext): Mismatch? {
        checkProperties()
        val existing = getExistingContent(d, ctx)
        val toContent = getToContent(d, ctx)
        val m = same(existing, toContent)
        if (m != null) {
            return m
        }
        doReplace(d, existing, getFromContent(d, ctx), toContent, ctx)
        return null
    }
}